### ISSUE TITLE
Client connection to node blocks node internal communication

### ISSUE THREAD

**Issue by Geod24** [bosagora/agora#312]
A bug that I discovered while working on the integration test: the transactions / block will only be gossiped *after* the connection to nodes are closed. This is probably because the re-entrancy of `vibe.web.rest` is not defined, so any network communication will just keep the object busy, which is terrible.
I'll need to come up with a simple example and report this upstream. Fortunately for us, the new RCP generator will fix this.

**Comment by Geod24** [comment]
I just witnessed this again: A node was retrieving a large amount of blocks (720) and the metric endpoint was unresponsive for few seconds (around 1m15s according to the logs).

**Comment by ferencdg** [comment]
I am wondering wether this unresponsiveness is caused by us instead of vibe.d. We are retrieving a maximum of 1024 blocks(720 in this case) in a sync call(as getBlocksFrom is not wrapped inside a task)

**Comment by ferencdg** [comment]
never mind, it is wrapped inside a task at a higher level

**Comment by Geod24** [comment]
It has recently come to my attention that `requestHTTP` [*is documented as synchronous*](https://vibed.org/api/vibe.http.client/requestHTTP).

**Comment by ferencdg** [comment]
do we have the logs somewhere? I also want to verify how long it takes to process 720 blocks...

**Comment by Geod24** [comment]
```
Jan 08 06:37:12 na-002 docker[22304]: [main(Ky6v) INF] Listening for requests on http://0.0.0.0:7826/
Jan 08 06:37:12 na-002 docker[22304]: 2021-01-08 06:37:12,175 Info [agora.network.NetworkManager] - Doing network discovery..
Jan 08 06:37:12 na-002 docker[22304]: [main(+O1h) INF] Listening for requests on http://[::]:9111/
Jan 08 06:37:12 na-002 docker[22304]: Failed to listen on 0.0.0.0:9111
Jan 08 06:37:12 na-002 docker[22304]: 2021-01-08 06:37:12,182 Info [agora.consensus.protocol.Nominator] - Starting nominating timer..
Jan 08 06:37:12 na-002 docker[22304]: 2021-01-08 06:37:12,183 Info [agora.network.NetworkManager] - Found new Validator: http://na-002.bosagora.io:6826 (key: GDNODE6ZXW2NNOOQIGN24MBEZRO5226LSMHGQA3MUAMYQSTJVR7XT6GH)
Jan 08 06:37:12 na-002 docker[22304]: 2021-01-08 06:37:12,196 Info [agora.network.NetworkManager] - Found new Validator: http://na-001.bosagora.io:5826 (key: GDNODE5T7TWJ2S4UQSTM7KDHU2HQHCJUXFYLPZDDYGXIBUAH3U3PJQC2)
Jan 08 06:37:12 na-002 docker[22304]: 2021-01-08 06:37:12,197 Info [agora.network.NetworkManager] - Found new Validator: http://na-001.bosagora.io:4826 (key: GDNODE4KTE7VQUHVBLXIGD7VEFY57X4XV547P72D37SDG7UEO7MWOSNY)
Jan 08 06:37:12 na-002 docker[22304]: 2021-01-08 06:37:12,347 Info [agora.network.NetworkManager] - Found new Validator: http://eu-002.bosagora.io:2826 (key: GDNODE2IMTDH7SZHXWDS24EZCMYCEJMRZWB3S4HLRIUP6UNGKVVFLVHQ)
Jan 08 06:37:12 na-002 docker[22304]: 2021-01-08 06:37:12,348 Info [agora.network.NetworkManager] - Found new Validator: http://eu-002.bosagora.io:3826 (key: GDNODE3EWQKF33TPK35DAQ3KXAYSOT4E4ACDOVJMDZQDVKP66IMJEACM)
Jan 08 06:37:13 na-002 docker[22304]: 2021-01-08 06:37:13,209 Info [agora.network.NetworkManager] - Found new Validator: http://na-002.bosagora.io:7826 (key: GDNODE7J5EUK7T6HLEO2FDUBWZEXVXHJO7C4AF5VZAKZENGQ4WR3IX2U)
Jan 08 06:37:14 na-002 docker[22304]: 2021-01-08 06:37:14,717 Info [agora.network.NetworkManager] - Retrieving blocks [1..720] from http://na-002.bosagora.io:6826..
Jan 08 06:37:15 na-002 docker[22304]: 2021-01-08 06:37:15,902 Info [agora.network.NetworkManager] - Received blocks [1..720]
Jan 08 06:38:30 na-002 docker[22304]: No response written for /favicon.ico
Jan 08 07:33:42 na-002 systemd[1]: Stopping Agora node 7...
```

As said, the logs don't show much. But the `favicon` things is something that shows up when a browser query the `/metrics` endpoint, which I did.

**Comment by Geod24** [comment]
`journalctl --unit=agora@7 --since=today` on `na-002` then `/1\.\.720`

**Comment by ferencdg** [comment]
are these the agora configs we are using in the docker?
  
  100ms for serializing 720 blocks might be a bit short... also 25 retries :)
  
  retry_delay: 50
  max_retries: 25
  timeout: 100

**Comment by ferencdg** [comment]
this might be explained by:

https://github.com/bpfkorea/agora/pull/1485
https://github.com/bpfkorea/agora/pull/1477

**Comment by Geod24** [comment]
I would have expected that transferring 720 blocks would leads to a few yield, wouldn't it ?

**Comment by ferencdg** [comment]
> I would have expected that transferring 720 blocks would leads to a few yield, wouldn't it ?

it probably yields as long as the receiver is faster than the sender, but I wonder what happens if we register our timers like this with Periodic.Yes:
`this.taskman.setTimer(2.seconds, &catchup, Periodic.Yes);` 

the timeout takes 3750 millisecs
(
retry_delay: 50
max_retries: 25
timeout: 100
)

and we put a call to the catchup into the scheduler queue every 2 seconds... I think after a while the queue will contain too many calls to 'catchup' and might starve other fibers... Changing the catchup interval to 20.seconds would likely make this problem less apparent.

Shouldn't we remove Periodic.Yes too? (I see the same pattern in other parts of our code btw)

**Comment by ferencdg** [comment]
> > I would have expected that transferring 720 blocks would leads to a few yield, wouldn't it ?
> 
> it probably yields as long as the receiver is faster than the sender, but I wonder what happens if we register our timers like this with Periodic.Yes:
> `this.taskman.setTimer(2.seconds, &catchup, Periodic.Yes);`
> 
> the timeout takes 3750 millisecs
> (
> retry_delay: 50
> max_retries: 25
> timeout: 100
> )
> 
> and we put a call to the catchup into the scheduler queue every 2 seconds... I think after a while the queue will contain too many calls to 'catchup' and might starve other fibers... Changing the catchup interval to 20.seconds would likely make this problem less apparent.
> 
> Shouldn't we remove Periodic.Yes too? (I see the same pattern in other parts of our code btw)

hmmm, actually it definitely yields, because the timeour is just 100msecs...

**Comment by ferencdg** [comment]
But let's just suppose this node for which we experienced hangup ran for 10 hours already... that means it already put 18000 of this 'cachup' calls into the queue from which it processed around 9000(a bit more than that), so 9000 is still in the queue. Then someone other fiber posts something into the queue, and he needs to wait: 9000*100ms = 15mins. We probably don't have to wait 100ms though cause the fiber will yield before that...

**Comment by Geod24** [comment]
Fixed in #1485 + some timeout adjustments.
